File: programmers/오픈채팅방.py
def solution(record):
    answer = []
    
    dic = {}
    result = []
    for r in record :
        s = r.split(" ")
        if len(s) == 3:
            order = s[0]
            id = s[1]
            name = s[2]
        else :
            order = s[0]
            id = s[1]
        if order == "Enter" :
            dic[id] = name
            result.append((1,id))

        elif order == "Leave":
            result.append((2,id))
        else :
            dic[id] = name

    for ord,i in result :
        str = dic[i]+"님이 "
        if ord == 1 :
            str += "들어왔습니다."
        else :
            str += "나갔습니다."
        answer.append(str)

    return answer

File: programmers/코테대비(2022.01.09)/오픈채팅방.java
import java.util.*;

class Solution {
    public String[] solution(String[] record) {
        ArrayList<String> arr = new ArrayList();
        ArrayList<String> arr2 = new ArrayList();
        Map<String,String> map = new HashMap<>();
        for(String str : record){

            String[] order = str.split(" ");
            if(order.length == 2){
                arr.add("님이 나갔습니다.");
                arr2.add(order[1]);
            }
            else if(order[0].equals("Enter")){
                map.put(order[1],order[2]);
                arr.add("님이 들어왔습니다.");
                arr2.add(order[1]);
            } else {
                map.put(order[1],order[2]);
            }

        }
        String[] answer = new String[arr.size()];
        for(int j=0;j<arr.size();j++){
            answer[j] = map.get(arr2.get(j))+arr.get(j);
        }
        
        return answer;
    }
}
